### PR TITLE
Port to Dune and random-generator's future

### PR DESCRIPTION
**Pull request by ostera**
Hello @gasche, and thanks for the work you've done on this lib :)

I was recently trying to use it and figured it'd be nicer if it was working directly with `dune` rather than with `ocamlbuild`. So here's a PR to begin that conversation.

A few things I'd like to talk about are:

1. Figure out unexported/unused combinators (`digit`, `lift`, `prod`)
2. Reformatting the code with `ocamlformat`
3. What do you have in mind for this lib's future?
4. Current documentation setup moved to `odoc`

What are your thoughts?

### CODE REVIEW

File: random-generator.install [hunk 1]
```diff
@@ -0,0 +1,19 @@
+lib: [
+  "_build/install/default/lib/random-generator/META" {"META"}
```
Comment (c-cube) [inline random-generator.install]: just a note: install files should not normally be in git. They're generated by dune when the library is built.

**Comment by gasche** [comment]
Yes, I was going to ask about why `.install` is included in your commit, as @c-cube mentions.

> I was recently trying to use it and figured it'd be nicer if it was working directly with dune rather than with ocamlbuild. So here's a PR to begin that conversation.

Fine with me, thanks for doing the work! I'm willing to merge the port-to-dune once we are satisfied with the PR state.

> Figure out unexported/unused combinators (digit, lift, prod)

Could you maybe document and expose `digit` and `prod`, and get rid of `lift`?

> Reformatting the code with ocamlformat

I've justed tried a run of `ocamlformat` and I don't like the result. One useful change is spacing around the infix operators in `shuffle`. The rest has some unobjectionable stuff but also changes that are hurting readability. I'm against a change with the current default option set of ocamlformat 0.7. Some examples of hurtful things:

```diff
-let map f gen = fun rand -> f (gen rand)
+let map f gen rand = f (gen rand)
```

```diff
 let guard p gen r =
-  match gen r with
-    | None -> None
-    | Some x ->
-      if p x then Some x else None
+  match gen r with None -> None | Some x -> if p x then Some x else None
```

```diff
-  let rec fix derec_gen param =
-    fun random fuel -> derec_gen (fix derec_gen) param random fuel
+  let rec fix derec_gen param random fuel =
+    derec_gen (fix derec_gen) param random fuel
```

```
 let binary gen1 gen2 merge =
   let open Fuel in
-  map'
-    (tick (prod split_int gen1 gen2))
-    (fun (v1, v2) -> merge v1 v2)
-
+  map' (tick (prod split_int gen1 gen2)) (fun (v1, v2) -> merge v1 v2)
```

> What do you have in mind for this lib's future?

I haven't used it recently, and I don't have any concrete plans to get it to evolve. I like the minimalism: it's doing one thing (combinators to generate values randomly) with a polished interface. I wouldn't want it to evolve into a Quickcheck (which has a language for generating and a language for testing). Nowadays I think that many uses of the library could be profitably replaced by [Crowbar](https://github.com/stedolan/crowbar). I'm happy to hear about your own needs/suggestions/ideas, or anyone's.

> Current documentation setup moved to odoc

I would be okay with that. (If I did the work myself I would try to keep an ocamldoc target as well, and I think it should be easy given how small the library is.)

File: random-generator.install (hunk 1 above)
Comment (ostera) [reply random-generator.install]: Thanks! I forgot to .gitignore this :)

**Comment by ostera** [comment]
> I haven't used it recently, and I don't have any concrete plans to get it to evolve. I like the minimalism: it's doing one thing (combinators to generate values randomly) with a polished interface. I wouldn't want it to evolve into a Quickcheck (which has a language for generating and a language for testing). Nowadays I think that many uses of the library could be profitably replaced by Crowbar. I'm happy to hear about your own needs/suggestions/ideas, or anyone's.

Mostly interested in seeing how well it works with BuckleScript. While I don't think any modifications to the code should be needed for it, it might require a functorization over `Random.State.t`. I'll let you know once I try that out.

**Comment by c-cube** [comment]
@gasche crowbar seems cool, but also kind of "high tech". I think there's still an important role for property-based testing without magic. However it could be based on quickcheck or hypothesis (cf [a note in rust's version of hypothesis](https://docs.rs/proptest/*/proptest/#differences-between-quickcheck-and-proptest) and [the original blog post](https://hypothesis.works/articles/integrated-shrinking/)) and in any case, the random monad offered by `random-generator` works well.

**Comment by gasche** [comment]
Hi from the moderately-distant past!

I wanted to do a small change on `random-generator` and I'm coming back to this PR first. I updated the current PR to squash @ostera's intermediate commits and fix the documentation generator -- still using ocamlbuild+ocamldoc, but in line with the new structure. I'm planning to migrate to dune+odoc, and update the PR again, and then merge.

Thanks!

**Comment by gasche** [comment]
In case anyone wonder: the change I would like to make on random-generator is to define it as a functor over a "base" random monad that just provides core primitives (mostly `Random.State`), to be able to instantiate it with a compute-the-whole-discrete-distribution monad to check that uniform algorithms really are uniform on small inputs. (My plan is to use `random-generator` in this way to test/debug code intended to land in QCheck.)
Incidentally this may also help with portability to non-standard runtimes.

**Comment by ostera** [comment]
Oh wow, this PR feels like it was ages ago. Sounds good to me 😄  Thanks for coming back to it 🙌🏽

**Comment by gasche** [comment]
### dune

I'm done working on the PR, and the result is a Dune build system which I like. The documentation generation is now done with `odoc`.

### automatic formatting

You mentioned `ocamlformat` earlier, and wanted to know what I thought of using it to enforce a style.

I just did a small experiment with various ocamlformat profiles, and my conclusion was that all those I could try produce output that is unpleasant enough that I don't want to use them. random-generator is a small amount of code mostly written by one person, and tolerating occasional style differences from a few external contributors sounds much more friendly than having to buy in the partially-obnoxious choices of a robot. So: no ocamlformat in this project for now.

Slightly more details: I have very few strong opinions on OCaml code formatting. I think my only strong opinion is that the formatting of "sequence" structures should be diff-oriented (each item on its own line to minimize diff noise). I haven't found an ocamlformat setup that works for this. Then `random-generator` is a bit harder than some other projects to format because I sometimes write functions that themselves return functions `let app f g = fun rand -> ...`, and I think it matters for readability because it is really a two-argument function returning a function, and (understandably) ocamlformat makes a mess of this. The overall result is that, except for a couple improvements in the placement of documentation comments, basically every change produced by running ocamlformat objectively makes things objectively worse, so why bother.

### future (testsuite, CI, etc.)

Two things are still missing by my current-day standards:
- a testsuite
- Continuous Integration (which probably needs a testsuite first)

One question for me to answer before I look at those is whether I (or other people of course) have any plan to use `random-generator` for anything in the future. In practice what I need when I write random generators these days is QCheck, with its support for shrinking (killer feature if you ask me), the driver, and generally all the important bits that are missing from here. But the QCheck generator interface is merely okay, not good -- but also "good enough" in practice for real projects. Making the QCheck generators more advanced or more minimal looks like it would run in boilerplate or compatibility issues (for example QCheck has the problem of having to duplicate everything in its Gen and Arbitrary module). I could also try to use random-generator generators on top of the QCheck machinery (I think it could work), but it's a lot of trouble for relatively small gains.

One concrete thing I would like to use random-generator for, as I mentioned, is as a way to debug random-generation algorithm by generating the full discrete distribution of the output for small inputs. I have a PR in preparation against QCheck that I could test that way. Incidentally this would also make it easy to write test for `random-generator`, which currently sounds tricky (how do we check that the output is "random enough"? fixing a seed and asserting some outputs is fragile and does not bring us much).

**Comment by c-cube** [comment]
Briefly :

- PR on qcheck sounds exciting. There's a lot of really cool work by Julien Debon that is cooking, to get integrated shrinking.
- I think there is a test in containers that samples some basic generators with a normal distribution and checks that the mean and variance are reasonable. Ironically it sometimes fails in CI. 

On May 1, 2021 4:28:40 PM EDT, Gabriel Scherer ***@***.***> wrote:
>### dune
>
>I'm done working on the PR, and the result is a Dune build system which
>I like. The documentation generation is now done with `odoc`.
>
>### automatic formatting
>
>You mentioned `ocamlformat` earlier, and wanted to know what I thought
>of using it to enforce a style.
>
>I just did a small experiment with various ocamlformat profiles, and my
>conclusion was that all those I could try produce output that is
>unpleasant enough that I don't want to use them. random-generator is a
>small amount of code mostly written by one person, and tolerating
>occasional style differences from a few external contributors sounds
>much more friendly than having to buy in the partially-obnoxious
>choices of a robot. So: no ocamlformat in this project for now.
>
>Slightly more details: I have very few strong opinions on OCaml code
>formatting. I think my only strong opinion is that the formatting of
>"sequence" structures should be diff-oriented (each item on its own
>line to minimize diff noise). I haven't found an ocamlformat setup that
>works for this. Then `random-generator` is a bit harder than some other
>projects to format because I sometimes write functions that themselves
>return functions `let app f g = fun rand -> ...`, and I think it
>matters for readability because it is really a two-argument function
>returning a function, and (understandably) ocamlformat makes a mess of
>this. The overall result is that, except for a couple improvements in
>the placement of documentation comments, basically every change
>produced by running ocamlformat objectively makes things objectively
>worse, so why bother.
>
>### future (testsuite, CI, etc.)
>
>Two things are still missing by my current-day standards:
>- a testsuite
>- Continuous Integration (which probably needs a testsuite first)
>
>One question for me to answer before I look at those is whether I (or
>other people of course) have any plan to use `random-generator` for
>anything in the future. In practice what I need when I write random
>generators these days is QCheck, with its support for shrinking (killer
>feature if you ask me), the driver, and generally all the important
>bits that are missing from here. But the QCheck generator interface is
>merely okay, not good -- but also "good enough" in practice for real
>projects. Making the QCheck generators more advanced or more minimal
>looks like it would run in boilerplate or compatibility issues (for
>example QCheck has the problem of having to duplicate everything in its
>Gen and Arbitrary module). I could also try to use random-generator
>generators on top of the QCheck machinery (I think it could work), but
>it's a lot of trouble for relatively small gains.
>
>One concrete thing I would like to use random-generator for, as I
>mentioned, is as a way to debug random-generation algorithm by
>generating the full discrete distribution of the output for small
>inputs. I have a PR in preparation against QCheck that I could test
>that way. Incidentally this would also make it easy to write test for
>`random-generator`, which currently sounds tricky (how do we check that
>the output is "random enough"? fixing a seed and asserting some outputs
>is fragile and does not bring us much).
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/gasche/random-generator/pull/5#issuecomment-830690084

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by gasche** [comment]
Note: I'm done turning `random-generator` into a functor over a probability monad (), and this let me experiment with fairness of generators as I expected. I still have to include a testsuite and some continuous integration to feel satisfied.

For me this repository remains an experimental device to play with random generation, rather than a library meant for production (I think that QCheck is better for that).

**Pull request merged**